odoo/osv/query.py

```diff
@@ -160,15 +160,47 @@ def add_joins_for_table(lhs):
                 from_clause.append(')')
                 add_joins_for_table(rhs)
 
-        for pos, table in enumerate(tables_to_process):
+        joined_aliases = [
+            join[0]
+            for joined_table in self.joins.values()
+            for join in joined_table]
+        pos = 0
+        for table in tables_to_process:
+            table_alias = get_alias_from_query(table)[1]
+            if table_alias in joined_aliases:
+                continue
             if pos > 0:
                 from_clause.append(',')
             from_clause.append(table)
-            table_alias = get_alias_from_query(table)[1]
+            pos += 1
             if table_alias in self.joins:
                 add_joins_for_table(table_alias)
 
         return "".join(from_clause), " AND ".join(self.where_clause), from_params + self.where_clause_params
 
     def __str__(self):
         return '<osv.Query: "SELECT ... FROM %s WHERE %s" with params: %r>' % self.get_sql()
+
+    def __iadd__(self, query):
+        """ Include another query into self """
+        self.where_clause += query.where_clause
+        self.where_clause_params += query.where_clause_params
+        for table in query.tables:
+            if table not in self.tables:
+                self.tables.append(table)
+        for join_table in query.joins:
+            self.joins.setdefault(join_table, [])
+            for join in query.joins[join_table]:
+                if join not in self.joins[join_table]:
+                    self.joins[join_table].append(join)
+        return self
+
+    def __add__(self, query):
+        """ Return a new copy of `self` that includes `query` """
+        result = Query(
+            list(self.tables or []),
+            list(self.where_clause or []),
+            list(self.where_clause_params or []),
+            deepcopy(self.joins or {}))
```

Comment:
if keys and values are immutable, `deepcopy` is probably not necessary

Reply:
I kept this deepcopy because the values are lists and thus mutable.